File: year2018/day04/python/solution.py
import re

def main():
  print(part_1())
  print(part_2())


def part_1():
  s_guard, s_total = 0, 0

  for guard, record in input.items():
    if sum(record) > s_total:
      s_guard, s_total = guard, sum(record)

  s_min, s_record = 0, input[s_guard]

  for minute in range(60):
    if s_record[minute] > s_record[s_min]:
      s_min = minute

  return s_guard * s_min


def part_2():
  s_min, s_freq, s_guard = 0, 0, 0

  for guard, record in input.items():
    for minute in range(60):
      if record[minute] > s_freq:
        s_min, s_freq, s_guard = minute, record[minute], guard

  return s_guard * s_min


def setup(path):
  global input

  with open(path, 'r') as file:
    input = setup_lines(sorted([row for row in file]))


def setup_lines(lines):
  records = {}
  currentGuard = 0
  sleepStart = 0

  for line in lines:
    match = re.match(r'^\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\] (?:Guard #(\d+) begins shift|(falls asleep)|(wakes up))$', line)
    if match:
      minute, guard, fallsAsleep, wakesUp = match.groups()
      if guard:
        currentGuard = int(guard)
        if currentGuard not in records:
          records[currentGuard] = [0] * 60
      if fallsAsleep:
        sleepStart = int(minute[-2:])
      if wakesUp:
        for i in range(sleepStart, int(minute[-2:])):
          records[currentGuard][i] = records[currentGuard][i] + 1 if records[currentGuard][i] else 1

  return records


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2018/day04/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: Map<number, number[]> = setup(file.toString().trim().split('\n').sort());
if(process.argv[2] === 'puzzle') {
    console.log(part_1());
    console.log(part_2());
}

export function part_1(): number {
    let [sGuard, sTotal] = [0, 0];
    for (let [guard, record] of input.entries()) if (record.reduce((x, y) => x + y, 0) > sTotal) [sGuard, sTotal]  = [guard, record.reduce((x, y) => x + y, 0)];
    let [sMin, sRecord] = [0, input.get(sGuard)!];
    for (let minute = 0; minute < 60; minute++) if (sRecord[minute] > sRecord[sMin]) sMin = minute;
    return sGuard * sMin;
}

export function part_2(): number {
    let [sMin, sFreq, sGuard] = [0, 0, 0];
    for (let [guard, record] of input.entries()) for (let minute = 0; minute < 60; minute++) if (record[minute] > sFreq) [sMin, sFreq, sGuard] = [minute, record[minute], guard];
    return sGuard * sMin;
}

function setup(lines: string[]): Map<number, number[]> {
    let [records, currentGuard, sleepStart] = [new Map<number, number[]>(), 0, 0];
    for (let line of lines) {
      const [_, minute, guard, fallsAsleep, wakesUp] = line.match(/^\[\d{4}-\d{2}-\d{2} \d{2}:(\d{2})\] (?:Guard #(\d+) begins shift|(falls asleep)|(wakes up))$/) || [];
      if (guard) {
        currentGuard = Number(guard);
        if (!records.has(currentGuard)) records.set(currentGuard, new Array(60).fill(0));
      }
      if (fallsAsleep) sleepStart = parseInt(minute, 10);
      if (wakesUp) for (let i = sleepStart; i < Number(minute); i++) records.get(currentGuard)[i] = (records.get(currentGuard)[i] || 0) + 1;
    }
    return records;
}